### PR TITLE
Pricing algorithms tests

### LINKED ISSUE #326: Tests for Pricing Algorithms
Add tests for price simulation, totals, taxes (net & gross)

### PR DESCRIPTION
**Pull request by Mikearaya**
Closes #326

### CODE REVIEW

**Comment by Mikearaya** [comment]
After spending a lot of hours trying to find a workaround on how to mock imported modules inside a file which all Adapters do, I was not able to find a solution that works. 
all the mocking https://jestjs.io/docs/jest-object#jestcreatemockfrommodulemodulename don't seem to work for EsModules.

the problem is all adapters use the base adapter functions imported in them to perform a certain operation

```
actions: ( params) {
...
    const pricingAdapter = DeliveryPricingAdapter.actions(params);
....

return {
      ...pricingAdapter,

      calculate: async () => {
        const taxRate = getTaxRate(context);
        DeliveryPricingAdapter.log(`DeliverySwissTax -> Tax Multiplicator: ${taxRate}`);
        params.calculationSheet.filterBy({ isTaxable: true }).forEach(({ isNetPrice, ...row }) => {
          if (!isNetPrice) {
            const taxAmount = row.amount - row.amount / (1 + taxRate);
            pricingAdapter.resultSheet().calculation.push({

```

as you can see on the above `pricingAdapter.calculationSheet.addTax` is used from the DeliveryPricingAdapter and that is what needs to be mocked.

I'll try to find a workaround but if anyone has an idea please share.

**Comment by pozylon** [comment]
Ok, let's dive into this:

Which file/function do you want to unit test?


Let's assume you wanted to create tests for delivery-swiss-tax.ts, looking at that file there is the following areas to cover:

1. getTaxRate
2. isDeliveryAddressInSwitzerland
3. calculate function in actions

The first 2 are straight forward, so let's look at 3 line by line:

```
        const taxRate = getTaxRate(context);
```

This is already unit tested


```
        DeliveryPricingAdapter.log(`DeliverySwissTax -> Tax Multiplicator: ${taxRate}`);
```

Logging doesn't need unit testing


```
        params.calculationSheet.filterBy({ isTaxable: true }).forEach(({ isNetPrice, ...row }) => {
```

calculationSheet.filterBy should be unit tested in "DeliveryPricingSheet.ts" in core-delivery.

```
          if (!isNetPrice) {
            const taxAmount = row.amount - row.amount / (1 + taxRate);
            pricingAdapter.resultSheet().calculation.push({
              ...row,
              amount: -taxAmount,
              isTaxable: false,
              isNetPrice: false,
              meta: { adapter: DeliverySwissTax.key },
            });
            pricingAdapter.resultSheet().addTax({
              amount: taxAmount,
              rate: taxRate,
              meta: { adapter: DeliverySwissTax.key },
            });
          } else {
            const taxAmount = row.amount * taxRate;
            pricingAdapter.resultSheet().addTax({
              amount: taxAmount,
              rate: taxRate,
              meta: { adapter: DeliverySwissTax.key },
            });
          }
        });
```
now the above is specific and needs a unit test if you aim for 100% coverage.


```
        return pricingAdapter.calculate();
```
calculate function of a pricing adapter is part of other unit tests




Ok so, how to write a unit test for 
```
          if (!isNetPrice) {
            const taxAmount = row.amount - row.amount / (1 + taxRate);
            pricingAdapter.resultSheet().calculation.push({
              ...row,
              amount: -taxAmount,
              isTaxable: false,
              isNetPrice: false,
              meta: { adapter: DeliverySwissTax.key },
            });
            pricingAdapter.resultSheet().addTax({
              amount: taxAmount,
              rate: taxRate,
              meta: { adapter: DeliverySwissTax.key },
            });
          } else {
            const taxAmount = row.amount * taxRate;
            pricingAdapter.resultSheet().addTax({
              amount: taxAmount,
              rate: taxRate,
              meta: { adapter: DeliverySwissTax.key },
            });
          }
        });
```

ways to do this:
1. mocking pricingAdapter
2. refactor: extract the code and inject the pricingAdapter.resultSheet()

mocking you have already tried so that leads to 2:

```
import { DeliveryPricingAdapter, DeliveryPricingDirector } from '@unchainedshop/core-delivery';
import {
  DeliveryPricingAdapterContext,
  DeliveryPricingCalculation,
  IDeliveryPricingAdapter,
  IDeliveryPricingSheet,
} from '@unchainedshop/types/delivery.pricing.js';

// https://www.ch.ch/de/mehrwertsteuersatz-schweiz/
export const SwissTaxCategories = {
  DEFAULT: {
    rate: () => {
      return 0.077;
    },
  },
  REDUCED: {
    value: 'reduced',
    rate: () => {
      return 0.025;
    },
  },
  SPECIAL: {
    value: 'special',
    rate: () => {
      return 0.037;
    },
  },
};

const getTaxRate = (context: DeliveryPricingAdapterContext) => {
  const taxCategoryFromProvider = context.provider?.configuration?.find(({ key }) => {
    if (key === 'swiss-tax-category') return true;
    return null;
  })?.value;

  if (taxCategoryFromProvider === SwissTaxCategories.REDUCED.value) {
    return SwissTaxCategories.REDUCED.rate();
  }
  if (taxCategoryFromProvider === SwissTaxCategories.SPECIAL.value) {
    return SwissTaxCategories.SPECIAL.rate();
  }
  return SwissTaxCategories.DEFAULT.rate();
};

const isDeliveryAddressInSwitzerland = ({ orderDelivery, order, country = null }) => {
  let countryCode = country?.toUpperCase().trim() || order.countryCode;

  const address = orderDelivery?.context?.address || order?.billingAddress;

  if (address?.countryCode > '') {
    countryCode = address.countryCode?.toUpperCase().trim();
  }

  return countryCode === 'CH' || countryCode === 'LI';
};

const applyTaxesWithRate = (
  resultSheet: IDeliveryPricingSheet,
  calculationItems: DeliveryPricingCalculation[],
  taxRate: number,
) => {
  calculationItems.forEach((calculation) => {
    const { isNetPrice, ...row } = calculation;
    if (!isNetPrice) {
      const taxAmount = calculation.amount - calculation.amount / (1 + taxRate);
      resultSheet.calculation.push({
        ...row,
        amount: -taxAmount,
        isTaxable: false,
        isNetPrice: false,
      });
      resultSheet.addTax({
        amount: taxAmount,
        rate: taxRate,
      });
    } else {
      const taxAmount = row.amount * taxRate;
      resultSheet.addTax({
        amount: taxAmount,
        rate: taxRate,
      });
    }
  });
};

export const DeliverySwissTax: IDeliveryPricingAdapter = {
  ...DeliveryPricingAdapter,

  key: 'shop.unchained.pricing.delivery-swiss-tax',
  version: '1.0.0',
  label: 'Apply Swiss Tax on Delivery Fees',
  orderIndex: 20,

  isActivatedFor: (context) => {
    return isDeliveryAddressInSwitzerland(context);
  },

  actions: (params) => {
    const pricingAdapter = DeliveryPricingAdapter.actions(params);
    const { context } = params;

    return {
      ...pricingAdapter,

      calculate: async () => {
        const taxRate = getTaxRate(context);
        DeliveryPricingAdapter.log(`DeliverySwissTax -> Tax Multiplicator: ${taxRate}`);

        const calculationItems = params.calculationSheet
          .filterBy({ isTaxable: true })
          .map((calculation) => ({
            ...calculation,
            meta: { adapter: DeliverySwissTax.key },
          }));

        applyTaxesWithRate(pricingAdapter.resultSheet(), calculationItems, taxRate);

        return pricingAdapter.calculate();
      },
    };
  },
};

DeliveryPricingDirector.registerAdapter(DeliverySwissTax);
```

Now you could easily mock resultSheet with something like { addTax: () ..., calculation: [] }. But mocking is a smell, so what could be the better option?

**Comment by pozylon** [comment]
It's important to note that in order to test if the plugin works as it should we also need an integration test that looks something like this:

1. add a plugin that adds 10 USD delivery fees to kitchensink
2. simulate adding a product to the cart that uses the fixed fee delivery plugin
3. check if delivery fee taxes are correctly calculated in graphql Order and OrderDelivery

**Comment by pozylon** [comment]
Now, let's assume you don't want to mock, what could we do?

```
import { DeliveryPricingAdapter, DeliveryPricingDirector } from '@unchainedshop/core-delivery';
import {
  DeliveryPricingAdapterContext,
  DeliveryPricingCalculation,
  IDeliveryPricingAdapter,
  IDeliveryPricingSheet,
} from '@unchainedshop/types/delivery.pricing.js';

// https://www.ch.ch/de/mehrwertsteuersatz-schweiz/
export const SwissTaxCategories = {
  DEFAULT: {
    rate: () => {
      return 0.077;
    },
  },
  REDUCED: {
    value: 'reduced',
    rate: () => {
      return 0.025;
    },
  },
  SPECIAL: {
    value: 'special',
    rate: () => {
      return 0.037;
    },
  },
};

const getTaxRate = (context: DeliveryPricingAdapterContext) => {
  const taxCategoryFromProvider = context.provider?.configuration?.find(({ key }) => {
    if (key === 'swiss-tax-category') return true;
    return null;
  })?.value;

  if (taxCategoryFromProvider === SwissTaxCategories.REDUCED.value) {
    return SwissTaxCategories.REDUCED.rate();
  }
  if (taxCategoryFromProvider === SwissTaxCategories.SPECIAL.value) {
    return SwissTaxCategories.SPECIAL.rate();
  }
  return SwissTaxCategories.DEFAULT.rate();
};

const isDeliveryAddressInSwitzerland = ({ orderDelivery, order, country = null }) => {
  let countryCode = country?.toUpperCase().trim() || order.countryCode;

  const address = orderDelivery?.context?.address || order?.billingAddress;

  if (address?.countryCode > '') {
    countryCode = address.countryCode?.toUpperCase().trim();
  }

  return countryCode === 'CH' || countryCode === 'LI';
};

const calculateTaxAmount = (amount, isNetPrice, taxRate) => {
  if (!isNetPrice) return amount - amount / (1 + taxRate);
  return amount * taxRate;
};

export const DeliverySwissTax: IDeliveryPricingAdapter = {
  ...DeliveryPricingAdapter,

  key: 'shop.unchained.pricing.delivery-swiss-tax',
  version: '1.0.0',
  label: 'Apply Swiss Tax on Delivery Fees',
  orderIndex: 20,

  isActivatedFor: (context) => {
    return isDeliveryAddressInSwitzerland(context);
  },

  actions: (params) => {
    const pricingAdapter = DeliveryPricingAdapter.actions(params);
    const { context } = params;

    return {
      ...pricingAdapter,

      calculate: async () => {
        const taxRate = getTaxRate(context);
        DeliveryPricingAdapter.log(`DeliverySwissTax -> Tax Multiplicator: ${taxRate}`);

        const calculationItems = params.calculationSheet
          .filterBy({ isTaxable: true })
          .map((calculation) => ({
            ...calculation,
          }));

        const resultSheet = pricingAdapter.resultSheet();

        calculationItems.forEach((calculation) => {
          const { isNetPrice, ...row } = calculation;

          const amount = calculateTaxAmount(calculation.amount, isNetPrice, taxRate);

          if (!isNetPrice) {
            resultSheet.calculation.push({
              ...row,
              amount: -amount,
              isTaxable: false,
              isNetPrice: false,
              meta: { adapter: DeliverySwissTax.key },
            });
          }
          resultSheet.addTax({
            amount,
            rate: taxRate,
            meta: { adapter: DeliverySwissTax.key },
          });
        });

        return pricingAdapter.calculate();
      },
    };
  },
};

DeliveryPricingDirector.registerAdapter(DeliverySwissTax);
```

with this refactored option you can now even test the remaining part where you calculate the tax amount isolated, so the remaining untested logic is this:

```
calculate: async () => {
        const taxRate = getTaxRate(context);
        DeliveryPricingAdapter.log(`DeliverySwissTax -> Tax Multiplicator: ${taxRate}`);

        const calculationItems = params.calculationSheet
          .filterBy({ isTaxable: true })
          .map((calculation) => ({
            ...calculation,
          }));

        const resultSheet = pricingAdapter.resultSheet();

        calculationItems.forEach((calculation) => {
          const { isNetPrice, ...row } = calculation;

          const amount = calculateTaxAmount(calculation.amount, isNetPrice, taxRate);

          if (!isNetPrice) {
            resultSheet.calculation.push({
              ...row,
              amount: -amount,
              isTaxable: false,
              isNetPrice: false,
              meta: { adapter: DeliverySwissTax.key },
            });
          }
          resultSheet.addTax({
            amount,
            rate: taxRate,
            meta: { adapter: DeliverySwissTax.key },
          });
        });

        return pricingAdapter.calculate();
      },
```

this only uses a bunch of functions and conditionals and is a case for an integration test, so there is no need to unit test the calculate of that plugin

**Comment by Mikearaya** [comment]
@pozylon i've added a unit test for all the pure functions used in the adapters and I think we can merge now. 
create another PR where we extract the utils into separate files and share between adapters and also refactor the adapters logic into more testable code like we discussed

File: packages/plugins/tests/free-payment.test.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import {  PaymentFreePrice } from '../src/pricing/free-payment';
+
+
+describe("PaymentFreePrice", () => {
```
Comment (pozylon) [inline packages/plugins/tests/free-payment.test.js:4]: I think it does not make sense to test a response of a no-argument function and dupplication of metadata. I mean when you change the properties for ex. increase the version this is going to fail and will lead to duplicate work.

I'd either:
- Remove the test file completely
- Actually test that free payment leads to 0.00 payment fee (= free) and not null/undefined/exceptions & remove the property test

File: packages/plugins/tests/discount-half-price-manual.test.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  it('Actions', async () => {
+    const context = {};
+    const actions = await HalfPriceManual.actions({ context });
+    expect(typeof actions.isValidForSystemTriggering).toBe('function');
+    expect(typeof actions.isValidForCodeTriggering).toBe('function');
+    expect(typeof actions.discountForPricingAdapterKey).toBe('function');
+  });
+
+  it('has correct properties', () => {
```
Comment (pozylon) [inline packages/plugins/tests/discount-half-price-manual.test.js:42]: remove property test

File: packages/plugins/tests/order-delivery.test.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import {  OrderDelivery } from '../src/pricing/order-delivery';
+
+describe("OrderDelivery ", () => {
+
+  it('isActivatedFor', () => {
+    expect(OrderDelivery.isActivatedFor()).toBeTruthy();
+  });
+
+  it('has correct keys', () => {
```
Comment (pozylon) [inline packages/plugins/tests/order-delivery.test.js:9]: dito

File: packages/plugins/tests/product-price-round.test.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import {configureProductPriceRounding, ProductPriceRoundSettings, roundToNext} from '../src/pricing/product-price-round'
+
+describe('roundToNext', () => {
```
Comment (pozylon) [inline packages/plugins/tests/product-price-round.test.js:3]: same here: split the tests up and call them:

roundToNext.test.js, configureProductPriceRounding.test.js

File: packages/plugins/tests/product-swiss-tax.test.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  it('REDUCED rate', () => {
+    expect(SwissTaxCategories.REDUCED.rate()).toBe(0.025);
+  });
+
+  it('SPECIAL rate', () => {
+    expect(SwissTaxCategories.SPECIAL.rate()).toBe(0.037);
+  });
+});
+
+describe('getTaxRate', () => {
```
Comment (pozylon) [inline packages/plugins/tests/product-swiss-tax.test.ts:20]: Move to getTaxRate.test.js

**Review by pozylon — CHANGES_REQUESTED** [review]
In general, the utils tests are great! about the plugins tests, filenames and directories you chose: most of the tests are testing util functions in the plugin package which makes total sense. so make sure that you move the unit tests as close to the actual files that they are testing.

usually where to put stuff:
**/file-to-test.test.js (unit tests belong next to file)
packages/tests/flow-to-test.test.js* (integration tests belong in tests folder) 
/e2e (end to end tests spreading over multiple systems -> kitchensink + admin ui for example)

**Comment by pozylon** [comment]
Almost there, cleanup and move some files, start the tests locally then merge at will

**Pull request merged**